### PR TITLE
Support for multiple variables files

### PR DESCRIPTION
**Pull request by jpotter**
Allows for multiple --variables options to be given

### CODE REVIEW

**Comment by kblomqvist** [comment]
Thanks for the PR. The logic seems to be correct. I need to figure out why the tests fail though.

**Comment by jpotter** [comment]
Thanks, Kim. Let me know what I can do -- I may have missed another place in the code that needs updating, I'll put a note on it for review.

**Comment by kblomqvist** [comment]
I think I have solved the problems with the failing tests (c405656690e8bf9c8c326b4cd0f3aa8a762be8a0).

However, as you also noted the dependency file list should be also update to be aware of multiple variable files. Also scons builder needs update to support this, https://github.com/kblomqvist/yasha/blob/master/yasha/scons.py.

Finally, I would like to have tests for this PR.

**Comment by kblomqvist** [comment]
I'm working with scons builder to make it not dependent on changes like this.

**Comment by kblomqvist** [comment]
Also in case variable file is not given but found by automatic lookup, it should be put into list.

So this
```python
 if not variables and not no_variable_file:
    for file in template_companion:
        if file.endswith(tuple(PARSERS.keys())):
            variables = click.open_file(file, "rb")
            break
```

should become this
```python
 if not variables and not no_variable_file:
    for file in template_companion:
        if file.endswith(tuple(PARSERS.keys())):
            variables = [click.open_file(file, "rb")] # <--
            break
```

**Comment by kblomqvist** [comment]
@jpotter Do you want to address the remaining issues?

1) Dependency list (`-M` or `-MD`) needs update as well (as you already pointed out)
2) In case variable file is not given but found by automatic lookup, the file should be passed forward as a list because variables will be iterated through later (cli.py:176)
3) Tests are needed to test that multiple variable files behave as expected

**Comment by jpotter** [comment]
@kblomqvist I've done #1 and #2 -- but will need more than a few minutes to do #3 and at the moment can't. (Tests are important though!) ...if you're okay waiting until I have time over the weekend, I can look into it then.

**Comment by kblomqvist** [comment]
Sure. Take your time. The current tests pass now.

**Comment by Hedanito** [comment]
Oh wow, I just was about to make the same pull request. I guess it won't be necessary anymore.

The only thing I'll add is that `click` returns multiple arguments as tuples, not lists. So instead of `variables = [click.open_file(file, "rb")]`, it should be `variables = (click.open_file(file, "rb"),)` to stay consistent.

**Comment by jpotter** [comment]
@Hedanito Ah, good catch; updated. I still need to write tests per @kblomqvist's request...

File: yasha/cli.py [hunk 1]
```diff
@@ -173,7 +173,7 @@ def cli(template_variables, template, output, variables, extensions, encoding, i
     if not variables and not no_variable_file:
         for file in template_companion:
             if file.endswith(tuple(PARSERS.keys())):
-                variables = [click.open_file(file, "rb")]
+                variables = (click.open_file(file, "rb"),)
```
[the remark this answers is not in the document]
Comment (jpotter) [reply yasha/cli.py]: Hmm -- I'd want to match your coding conventions, however I thought convention for single-item tuples to include the comma? This avoids the situation where the tuple definition gets parsed as parens.

```
>>> print ( ('hello') == 'hello' )
True
>>> print ( ('hello,') == 'hello' )
False
```

[the remark this answers is not in the document]
Comment (jpotter) [reply yasha/cli.py]: (Copy-paste error on my first comment; edited it right after I posted it.)

[the remark this answers is not in the document]
Comment (kblomqvist) [reply yasha/cli.py]: Oops my mistake. I confused with the line 175. In my example `tuple(item)`, the item should be a list. However, in this case it's not a list like in line 175.

**Comment by kblomqvist** [comment]
The test does not need to be complex. Something like this is enough.

```python
def test_two_explicit_variable_files(tmpdir):
    # Template to calculate a + b 
    tpl = tmpdir.join('template.j2')
    tpl.write('{{ a + b }}')

    # First variable file
    a = tmpdir.join('a.yaml')
    a.write('a: 1')

    # Second variable file
    b = tmpdir.join('b.toml')
    b.write('b = 2')

    errno = call(('yasha', '-v', str(a), '-v', str(b), str(tpl)))
    assert errno == 0

    output = tmpdir.join('template')
    assert output.read() == '3' # a + b = 1 + 2 = 3
```

**Comment by jpotter** [comment]
@kblomqvist Thanks Kim. I've been having problems getting the full set of tests to run (peaking in the .travis.yml to see how they were triggered -- test_build_automation_c, test_stdin_and_out, and test_subprocess_with_unknown_cmd aren't passing on my OS X laptop). pytest is showing the test you suggested as passing though.

I updated the test to handled the case where variables are redefined between two variables files, on the basis that this should at least be defined / deterministic.

Let me know if this looks ok?

**Comment by kblomqvist** [comment]
@jpotter This should be now part of the version 4.2 available via pypi. Thanks.

**Comment by jpotter** [comment]
Thanks @kblomqvist!

**Pull request merged**